OOP school library: add Person, Student and Teacher classes

VelzckC0D3 opened the pull request:
# ✨ Requested Implementations ✨

### ✔️ _Created a Person class with the following requirements_

>- _Instance vars: @id, @name, and @age._
>- _Constructor with name, age, and parent_permission as parameter. name and parent_permission are optional and have default >values of "Unknown" and true._
>- _Getters for @id, @name, and @age._
>- _Setters for @name and @age._
>- _Private method of_age? that returns true if @age is greater or equal to 18 and false otherwise._
>- _Public method can_use_services? that returns true if person is of age or if they have permission from parents._

### ✔️ _Created a Student class with the following requirements_
>- _Inherits from Person._
>- _Constructor extends parent's constructor by adding @classroom and a parameter for it._
>- _Method play_hooky that returns "¯\(ツ)/¯"._

### ✔️ _Created a Teacher class with the following requirements_
>- _Inherits from Person._
>- _Constructor extends parent's constructor by adding @specialization and a parameter for it._
>- _Override can_use_services? so it always returns true._

# 💡 Thank you for your review! 💡

VelzckC0D3 committed:
Class Person

- Added a Class Person with the implementation of all the requirements & some test cases

VelzckC0D3 committed:
Class Student

- Added a Student Class with the implementation of all the requirements

VelzckC0D3 committed:
Class Teacher

- Added a Teacher class with the implementation of all the requirements

VelzckC0D3 committed:
Linters

Linter correction

Shedrack-Sunday reviewed (changes requested):
Hi @VelzckC0D3 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

# Invalid Code Review Request 🔴 

You have submitted a project, that does not have a personalized README file

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

VelzckC0D3 committed:
Update README.md

- Added a personalized Readme file

omar-muhamad commented on person.rb:
```diff
@@ -0,0 +1,36 @@
+    @parent_permission = parent_permission
+  end
+
+  attr_reader :id
+  attr_accessor :age, :name
```
- [x] [Optional] It would be better professionally if you move the attributes at the top of the class even before the initialize method.

VelzckC0D3 replied:
Done! thank you for the suggestion

omar-muhamad commented on person.rb:
```diff
@@ -0,0 +1,36 @@
+    @parent_permission || of_age?
+  end
+
+  # TESTING PURPOSES #
+  def permission
+    can_use_services? ? 'Allowed' : 'Not allowed'
+  end
+
+  def showcase
+    puts "My name is #{@name} and I am #{@age} years old, I'm #{permission} to use the services"
+  end
+  # TESTING PURPOSES #
+
+  private
+
+  def of_age?
+    @age >= 18
+  end
+end
+
+# TESTING PURPOSES #
+test1 = Person.new(18, 'John', parent_permission: false)
+test1.showcase
+# TESTING PURPOSES #
```
- [x] Your code is impeccable, but it is considered bad practice to leave testing code among your code itself. If you wanna test it professionally, Please make a new file for testing and you can try whatever you want in it `or` delete the testing code after you are sure that your code is working just fine.

omar-muhamad commented on student.rb:
```diff
@@ -0,0 +1,12 @@
+require './person'
+
+class Student < Person
+  def initialize(age, classroom, _name = 'Unknown', parent_permission: true)
+    super(age, 'Unknown', parent_permission)
+    @classroom = classroom
+  end
```
- [x] For the inheritance of the `name` variable, you should put the variable itself in `super`, not the default value of it. Please fix that.
[Hint]: It will be like that:
```ruby
def initialize(age, classroom, name = 'Unknown', parent_permission: true)
    super(age, name, parent_permission)
    @classroom = classroom
 end
```

omar-muhamad reviewed (changes requested):
# Changes Required ♻️
### Hi @VelzckC0D3,

### Nice job, you are almost there.

![giphy (2)](https://user-images.githubusercontent.com/68796232/223284713-d4ea97f0-5ad2-438b-8ac5-1481a991da3f.gif)

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- [ ] Please check the inline comments below ⬇️.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

VelzckC0D3 committed:
Required Changes

- Deleted test cases since it's not a best practice
- Corrected the inheritance of the name variable in the Student component

VelzckC0D3 committed:
Merge branch 'person-student-teacher' of https://github.com/VelzckC0D3/School_Library into person-student-teacher

VelzckC0D3 committed:
Linters check

- Corrected linter error

Davidosky007 reviewed (approved):
Hi @VelzckC0D3 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
*  Classes and Objects in Ruby implemented properly 👍 
* Encapsulation and Inheritance implemented with Ruby properly 👍 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

VelzckC0D3 committed:
minimal correction
